src/Conversion/ONNXToKrnl/Tensor/SizeOf.cpp

```diff
@@ -0,0 +1,57 @@
+//===---------------- SizeOf.cpp - Lowering Gather_nd Op
+//-------------------===//
+//
+// Copyright 2019 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Size Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+
+using namespace mlir;
+
+struct ONNXSizeOpLowering : public ConversionPattern {
+  ONNXSizeOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXSizeOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    // Gather info.
+    Location loc = op->getLoc();
+    Value alloc;
+    bool insertDealloc = checkInsertDealloc(op);
+    ONNXSizeOp sizeOp = llvm::dyn_cast<ONNXSizeOp>(op);
+
+    ONNXSizeOpAdaptor operandAdaptor(operands);
+    Value data = operandAdaptor.data();
+    ArrayRef<int64_t> dataShape = data.getType().cast<MemRefType>().getShape();
+    Value resultOperand = sizeOp.size();
+    ValueRange indices;
+    MemRefType memRefType = convertToMemRefType(*op->result_type_begin());
+
+    if (hasAllConstantDimensions(memRefType))
+      alloc = insertAllocAndDealloc(memRefType, loc, rewriter, insertDealloc);
+    else
+      alloc = insertAllocAndDealloc(
+          memRefType, loc, rewriter, insertDealloc, {resultOperand});
+
+    // The minimum number of elements in the input tensor has to be 1.
+    unsigned noElements = 1;
+    for (unsigned i = 0; i < dataShape.size(); i++)
+      noElements *= dataShape[i];
```

Comment:
Could you please handle the situation where dimensions in the `data`'s shape are unknown? E.g. `tensor<?x?xf32>`. In that situation, `dataShape[i] = -1` and we need to get the dimension via DimOp.